### PR TITLE
Add tests covering complex use of window functions

### PR DESCRIPTION
**Pull request by ebd2**
In preparation for work on window function performance, add tests covering more complex use of window functions.

I think this covers most of what presto implements. Notable omissions in what presto implements that are not covered:

- named window definitions
- RANGE with n PREVIOUS/FOLLOWING.

### CODE REVIEW

**Comment by ebd2** [comment]
@mattsfuller: would appreciate a look since you know the spec relatively well
@maciejgrzybek: Since you'll be looking at this too, it'd be nice to get your input.

If anybody has an idea of how to make this easier to review and, in particular, how to make the results more obviously correct, that would be helpful.

I'll go back through for query formatting tomorrow.

**Comment by ebd2** [comment]
Closing PR against prestodb/presto while we do internal review. Sorry about that.

**Comment by ebd2** [comment]
@maciejgrzybek has reviewed internally. Reopening.

**Comment by ebd2** [comment]
Rebased onto origin/master. @electrum: would you mind taking a look since you've looked over the other window function stuff I've done recently?

I've left the commits unsquashed in case you wanted to see how things evolved. If you prefer, I'll squash 08ff1c2 and e3c3384 into 9dea269. I was planning on leaving the unit tests in a separate commit.

**Comment by electrum** [comment]
@ebd2 can you squash into the final commits and I'll review and merge?

**Comment by ebd2** [comment]
@electrum: Squashed and rebased. Re-ran unit and product tests. Back to you.

**Comment by ebd2** [comment]
@electrum: rebased.

File: presto-main/src/test/java/com/facebook/presto/operator/window/TestMultipleWindowSpecifications.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.window;
+
+        import org.testng.annotations.Test;
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/operator/window/TestMultipleWindowSpecifications.java]: Indentation

**Comment by electrum** [comment]
I verified the queries in `TestMultipleWindowSpecifications` against PostgreSQL.

File: presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/rowsRangeLeadLag.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+-- database: presto; groups: window;
+select orderkey, suppkey,
+discount,
+lead(discount) over (partition by suppkey order by orderkey desc) next_discount,
+extendedprice,
+lag(extendedprice) over (partition by discount order by extendedprice range current row) previous_extendedprice
+
```
Comment (electrum) [inline presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/rowsRangeLeadLag.sql]: Extra newlines

File: presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/noOrderAllRowsPeers.sql [hunk 3]
```diff
@@ -0,0 +1,2 @@
+-- database: presto; groups: window;
+select orderkey, suppkey, discount, rank() over (partition by suppkey) from tpch.tiny.lineitem where partkey = 272
```
Comment (electrum) [inline presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/noOrderAllRowsPeers.sql]: This should have newlines to match the others

**Comment by electrum** [comment]
Merged, thanks!

**Pull request closed without merging**